timemachine/cpp/src/nonbonded.hpp

```diff
@@ -1,102 +1,32 @@
 #pragma once
 
-#include "neighborlist.hpp"
+#include "nonbonded_dense.hpp"
+#include "nonbonded_pairs.hpp"
 #include "potential.hpp"
-#include "vendored/jitify.hpp"
-#include <array>
 #include <vector>
 
 namespace timemachine {
 
-typedef void (*k_nonbonded_fn)(
-    const int N,
-    const double *__restrict__ coords,
-    const double *__restrict__ params, // [N]
-    const double *__restrict__ box,
-    const double *__restrict__ dl_dp,
-    const double *__restrict__ coords_w, // 4D coords
-    const double *__restrict__ dw_dl,    // 4D derivatives
-    const double beta,
-    const double cutoff,
-    const int *__restrict__ ixn_tiles,
-    const unsigned int *__restrict__ ixn_atoms,
-    unsigned long long *__restrict__ du_dx,
-    unsigned long long *__restrict__ du_dp,
-    unsigned long long *__restrict__ du_dl_buffer,
-    unsigned long long *__restrict__ u_buffer);
-
 template <typename RealType, bool Interpolated> class Nonbonded : public Potential {
 
 private:
-    std::array<k_nonbonded_fn, 16> kernel_ptrs_;
-
-    int *d_exclusion_idxs_; // [E,2]
-    double *d_scales_;      // [E, 2]
-    int *d_lambda_plane_idxs_;
-    int *d_lambda_offset_idxs_;
-    int *p_ixn_count_; // pinned memory
-
-    double beta_;
-    double cutoff_;
-    Neighborlist<RealType> nblist_;
-
-    const int E_;
-    const int N_;
-
-    double nblist_padding_;
-    double *d_nblist_x_;    // coords which were used to compute the nblist
-    double *d_nblist_box_;  // box which was used to rebuild the nblist
-    int *d_rebuild_nblist_; // whether or not we have to rebuild the nblist
-    int *p_rebuild_nblist_; // pinned
-
-    unsigned int *d_perm_; // hilbert curve permutation
-
-    double *d_w_;     //
-    double *d_dw_dl_; //
-
-    double *d_sorted_x_;     //
-    double *d_sorted_w_;     //
-    double *d_sorted_dw_dl_; //
-    double *d_sorted_p_;     //
-    double *d_unsorted_p_;   //
-    double *d_sorted_dp_dl_;
-    double *d_unsorted_dp_dl_;
-    unsigned long long *d_sorted_du_dx_; //
-    unsigned long long *d_sorted_du_dp_; //
-    unsigned long long *d_du_dp_buffer_; //
-
-    unsigned int *d_bin_to_idx_;
-    unsigned int *d_sort_keys_in_;
-    unsigned int *d_sort_keys_out_;
-    unsigned int *d_sort_vals_in_;
-    unsigned int *d_sort_storage_;
-    size_t d_sort_storage_bytes_;
-
-    bool disable_hilbert_;
-
-    void hilbert_sort(const double *d_x, const double *d_box, cudaStream_t stream);
-
-    jitify::JitCache kernel_cache_;
-    jitify::KernelInstantiation compute_w_coords_instance_;
-    jitify::KernelInstantiation compute_permute_interpolated_;
-    jitify::KernelInstantiation compute_add_du_dp_interpolated_;
+    NonbondedDense<RealType, Interpolated> dense_;
+    NonbondedPairs<RealType, true, Interpolated> exclusions_; // implement exclusions as negated NonbondedPairs
```

Comment:
Subjective nit
```suggestion
    bool Negated = true;
    NonbondedPairs<RealType, Negated, Interpolated> exclusions_;
```

Reply:
I like it, added in c299d45. (needed `static const` for this to compile)